packages/gatsby/src/utils/webpack.config.js

```diff
@@ -71,21 +71,11 @@ module.exports = async (
   }
 
   function getHmrPath() {
-    let hmrBasePath = `${program.ssl ? `https` : `http`}://${
-      program.host
-    }:${webpackPort}/`
-
+    // ref: https://github.com/gatsbyjs/gatsby/issues/8348
+    // construct the webpack url relative to where the request is coming from (if public path is not set)
+    const hmrBasePath = process.env.GATSBY_WEBPACK_PUBLICPATH || `/`
     const hmrSuffix = `__webpack_hmr&reload=true&overlay=false`
 
-    if (process.env.GATSBY_WEBPACK_PUBLICPATH) {
-      const pubPath = process.env.GATSBY_WEBPACK_PUBLICPATH
-      if (pubPath.substr(-1) === `/`) {
-        hmrBasePath = pubPath
-      } else {
-        hmrBasePath = `${pubPath}/`
-      }
-    }
-
```

Comment:
please don't remove that part - we still want to ensure path ends with `/` I think

Reply:
@pieh updated PR with the changes. I tested with following commands
- `GATSBY_WEBPACK_PUBLICPATH=/ gatsby develop -H 0.0.0.0` - setting the public path
- `gatsby develop -H 0.0.0.0` - without setting the public path

In both the cases, I was able to access the gatsby developement server running inside my vagrant machine

Reply:
@pieh @drubin I guess there is no difference between first and second command when it comes to testing this piece of code as thing are falling back to `/` anyway. One think that was not tested is testing with nginx as reverse proxy. I don't have that setup. In any case, what exactly to test with an nginx setup? let's say `mysite.local/gatsby` is proxied to 'localhost:8000', then what should be set in public path?

Reply:
@drubin Thanks for the clarification. As of now, I have not tested with an nginx setup. I will update here in case I'm able to test with an nginx setup.

Reply:
if you need to set custom `GATSBY_WEBPACK_PUBLICPATH` to point to `/gatsby` then there is no sense in testing this as it wouldn't work with current setup as well

Reply:
@pieh we can test if the ending `/` part and the default `GATSBY_WEBPACK_PUBLICPATH` takes effect instead of fallback relative path `/` I guess? I'm not sure how else this part can be tested. I don't find any relevant test cases for this piece of code too. But I'm not familiar with codebase and might have missed something.

Reply:
> Do you mean existing logic with manual url construction also might not work when public path set to `/gatsby`?

Yes - right now we construct url to be `http(s)://host:port/` so if user visits via `mysite.local/gatsby` then this wouldn't work and user would need to set `GATSBY_WEBPACK_PUBLICPATH` anyway, and this PR doesn't remove `GATSBY_WEBPACK_PUBLICPATH` so nothing changes with this PR